.travis.yml

```diff
@@ -85,6 +80,7 @@ install:
 script:
   - if [[ $TEST_COVERAGE == 'true' ]]; then composer test-coverage ; else composer test ; fi
```

Comment:
We should use docker here too. We need to install xdebug on the docker image and then we can run tests with coverage:

```
$ docker-compose run --rm php ./vendor/bin/phpunit --colors=always --coverage-clover clover.xml
```

or, for test only:

```
$ docker-compose run --rm php ./vendor/bin/phpunit --colors=always
```

Reply:
Coverage is only tested on one matrix element.  The output of running coverage on Docker would not be used.  I don't think Docker coverage is necessary.  Is there a place the output would be used I'm not seeing?

Reply:
@TomHAnderson We need to find a way to load xdebug extension only for coverage test.

I think should be better to create a `.docker` directory to put there docker files as @webimpress suggested.

Then we could:
- Write the entry point file and use `COPY` or `ADD` in the `Dockerfile`, without to write it "on the fly" building the image
- Install xdebug extension, without enabling it
- Create sh commands and copy them in the Dockerfile image

Then we can use a test-with-coverage script that will enable xdebug first:

```bash
#!/bin/sh

docker-php-ext-enable xdebug \
&& ./vendor/bin/phpunit --colors=always --coverage-clover clover.xml
```

What do you think?

Reply:
@TomHAnderson My idea is that if we are implementing docker, we should use it for everything.

It's not just a coverage, but are tests too. Right now tests are executed 2 times. First in the travis-ci environment (that it's different from the container) with coverage, then from the container.
 
I think we should use every command from the container, even `composer install` and `php-coveralls`.

After all these changes, we don't need to do anything on travis-ci environment, and then we can use a simple image on travis.

It's just my opinion and a suggestion :)

Reply:
I understand that, thanks.  

In this same thread @Ocramius suggests doing inside (Docker) and out (Travis) testing: https://github.com/zfcampus/zf-apigility-doctrine/pull/308#issuecomment-378137276

@webimpress I need sign-off from you too before I'm going to go with a Docker-only approach to unit tests.

Reply:
I get the reasoning from @Ocramius , and generally agree. The idea should be to run the tests via docker _always_, for every environment, and not using the travis-ci environment; the rationale is to ensure that the CI environment mimics the environment maintainers use.

That said: only do it if it can be feasibly setup. If this is going to take you days to accomplish, then I'm not 100% convinced it's worth it; in that case, the docker setup is simply so contributors do not need to have the required services (mongo, RDBMS systems, etc) running on their own machine, nor need to configure a `phpunit.xml` with the specifics of their own services.

If it _is_ possible to do easily, let's do it. Use build arguments to do things like enable xdebug and indicate test coverage should be run.